How do we specify a 3 segment Assembly Version (i.e. 0.4.1)?

I am attempting to migrate an existing project to Nerdbank.GitVersioning. The thought put into how this works in CI environments is great, which is why we would like to use it on this project.

However, we are running into trouble because there doesn't appear to be a way to specify a 3-segment AssemblyVersion number, which is what the existing version specifies.

## Existing Version from NuGet

PackageVersion: 0.4.1
```c#
[assembly: AssemblyVersion("0.4.1.0")]
[assembly: AssemblyInformationalVersion("0.4.1.0 commit:[15dc7a18b9]")]
```

## Patch Version

Since this is a port of 0.4.1, we cannot change that part of the version number without causing user confusion (unless of course we also port over the changes from 0.4.2 or next higher version that exists). So, the patch should be versioned as follows.

PackageVersion: 0.4.1.1
```c#
[assembly: AssemblyVersion("0.4.1.0")] // Must remain the same for binary compatibility on .NET Framework
[assembly: AssemblyInformationalVersion("0.4.1.1 commit:[xxxxxxxxxx]")]
[assembly: AssemblyFileVersion("0.4.1.1")]
```

## What we tried

First I tried setting the `<GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>`, but that doesn't have any effect on the output of Nerdbank.GitVersioning.

I looked at the [schema of version.json](https://raw.githubusercontent.com/AArnott/Nerdbank.GitVersioning/master/src/NerdBank.GitVersioning/version.schema.json) to see if there was an option to specify it explicitly and found `"assemblyVersion"` can be. But there is a problem: It defines the version as `major.minor`, so when I specify `"0.4.1"` it shows a warning and the AssemblyVersion is generated as `0.4.0.0` rather than the expected `0.4.1.0`.

Short of completely disabling Nerdbank.GitVersioning for our release build and using a custom script, is there any option to specify `0.4.1` as the AssemblyVersion?

I attempted both with 3.4.344 and 3.5.54-alpha, but got the same result each time.

```c#
[assembly: AssemblyFileVersion("0.4.1.1")]
[assembly: AssemblyInformationalVersion("0.4.1+5482e8d56c")]
[assembly: AssemblyVersion("0.4.0.0")]
```

The `version.json` file was committed as follows before the build:

```json
{
  "$schema": "https://raw.githubusercontent.com/AArnott/Nerdbank.GitVersioning/master/src/NerdBank.GitVersioning/version.schema.json",
  "version": "0.4.1.1",
  "assemblyVersion": {
    "version": "0.4.1"
  },
  "versionHeightOffset": "0",
  "publicReleaseRefSpec": [
    "^refs/heads/main$",
    "^refs/heads/master$",
    "^refs/heads/release/v\\d+\\.\\d+(?:\\.\\d+)?(?:\\.\\d+)?$",
    "^refs/tags/v\\d+\\.\\d+"
  ],
  "cloudBuild": {
    "buildNumber": {
      "enabled": true
    }
  },
  "release": {
    "branchName": "release/v{version}"
  }
}
```

I made a second attempt with 3.4.253-gac36c2b649 (from Azure DevOps) setting the precision to "build" and it correctly output AssemblyVersion 0.4.1.

```json
{
  "$schema": "https://raw.githubusercontent.com/AArnott/Nerdbank.GitVersioning/master/src/NerdBank.GitVersioning/version.schema.json",
  "version": "0.4.1.1",
  "assemblyVersion": {
    "version": "0.4.1",
    "precision": "build"
  },
  "versionHeightOffset": "0",
  "publicReleaseRefSpec": [
    "^refs/heads/main$",
    "^refs/heads/master$",
    "^refs/heads/release/v\\d+\\.\\d+(?:\\.\\d+)?(?:\\.\\d+)?$",
    "^refs/tags/v\\d+\\.\\d+"
  ],
  "cloudBuild": {
    "buildNumber": {
      "enabled": true
    }
  },
  "release": {
    "branchName": "release/v{version}"
  }
}
```

However, since it isn't possible to tell if that is due to the fact that it was simply ignoring `version` and using `precision`, I made an attempt to bump the version to the next real-world version of 0.5. It produces AssemblyVersion of 0.4.1, but the version number is 0.5.1.46059 instead of 0.5.

```json
{
  "$schema": "https://raw.githubusercontent.com/AArnott/Nerdbank.GitVersioning/master/src/NerdBank.GitVersioning/version.schema.json",
  "version": "0.5",
  "assemblyVersion": {
    "version": "0.4.1",
    "precision": "build"
  },
  "versionHeightOffset": "0",
  "publicReleaseRefSpec": [
    "^refs/heads/main$",
    "^refs/heads/master$",
    "^refs/heads/release/v\\d+\\.\\d+(?:\\.\\d+)?(?:\\.\\d+)?$",
    "^refs/tags/v\\d+\\.\\d+"
  ],
  "cloudBuild": {
    "buildNumber": {
      "enabled": true
    }
  },
  "release": {
    "branchName": "release/v{version}"
  }
}
```

```c#
[assembly: AssemblyFileVersion("0.5.1.46059")]
[assembly: AssemblyInformationalVersion("0.5.1+ebb31694c8")]
[assembly: AssemblyVersion("0.4.1.0")]
```

I then attempted to use 0.5.0 as the version number instead of 0.5, and that made the assembly version revert to 0.4.0 (not to mention, it didn't set the version at 0.5.0 either, but the informational version is what I had hoped for).

```json
{
  "$schema": "https://raw.githubusercontent.com/AArnott/Nerdbank.GitVersioning/master/src/NerdBank.GitVersioning/version.schema.json",
  "version": "0.5.0",
  "assemblyVersion": {
    "version": "0.4.1",
    "precision": "build"
  },
  "versionHeightOffset": "0",
  "publicReleaseRefSpec": [
    "^refs/heads/main$",
    "^refs/heads/master$",
    "^refs/heads/release/v\\d+\\.\\d+(?:\\.\\d+)?(?:\\.\\d+)?$",
    "^refs/tags/v\\d+\\.\\d+"
  ],
  "cloudBuild": {
    "buildNumber": {
      "enabled": true
    }
  },
  "release": {
    "branchName": "release/v{version}"
  }
}
```

```c#
[assembly: AssemblyFileVersion("0.5.0.1")]
[assembly: AssemblyInformationalVersion("0.5.0+3e982c4b05")]
[assembly: AssemblyVersion("0.4.0.0")]
```

Then I tried using `0.5.{height}` for the version number to see if I could use `versionHeightOffset` to wrangle `{height}` back to 0, but that generates an exception when trying to build:

```
F:\Users\shad\.nuget\packages\nerdbank.gitversioning\3.4.253-gac36c2b649\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: System.FormatException: Failure while reading version.json from commit df155680ba35f6008feea44574bcbdd542fe7bc6. Fix this commit with rebase if this is an error, or review this doc on how to migrate to Nerdbank.GitVersioning: https://github.com/dotnet/Nerdbank.GitVersioning/blob/master/doc/migrating.md ---> System.FormatException: The value "0.5.{height}" is not a valid semantic version.
```

Unfortunately, there are some strict rules in place that expect us to use semantic versioning. This is fine when you can use semantic versioning, but when porting code that needs to be version aligned and still needs to be patched in .NET to fix our mistakes, that doesn't seem to be possible.

While these are logical defaults, it would be great if this tool could:

1. Provide an option to turn off semantic versioning validation
2. Provide an option to pass in the 4 version arguments (package version, assembly version, file version, and informational version) to the nbgv tool as needed for the release, while calculating the omitted arguments as usual

All I am saying is that when using CI and dev it makes perfect sense to make the versions based on commit hashes rather than auto-incremented values that vary per environment. But when doing a release we need ultimate control over these values to prevent a versioning issue from blocking the release, such as aligning the versions to match some historical convention of the software we are releasing or to align the version to match an existing version we are porting.

--- a/src/NerdBank.GitVersioning.Tests/BuildIntegrationTests.cs
+++ b/src/NerdBank.GitVersioning.Tests/BuildIntegrationTests.cs
@@ -1009,13 +1009,19 @@ public abstract class BuildIntegrationTests : RepoTestBase, IClassFixture<MSBuil
     {
         // Function should be very similar to VersionOracle.GetAssemblyVersion()
         var assemblyVersion = (versionOptions?.AssemblyVersion?.Version ?? versionOptions.Version.Version).EnsureNonNegativeComponents();
-        var precision = versionOptions?.AssemblyVersion?.Precision ?? VersionOptions.DefaultVersionPrecision;
 
-        assemblyVersion = new System.Version(
-            assemblyVersion.Major,
-            precision >= VersionOptions.VersionPrecision.Minor ? assemblyVersion.Minor : 0,
-            precision >= VersionOptions.VersionPrecision.Build ? version.Build : 0,
-            precision >= VersionOptions.VersionPrecision.Revision ? version.Revision : 0);
+        if (versionOptions?.AssemblyVersion?.Version is null)
+        {
+            VersionOptions.VersionPrecision precision = versionOptions?.AssemblyVersion?.Precision ?? VersionOptions.DefaultVersionPrecision;
+            assemblyVersion = version;
+
+            assemblyVersion = new Version(
+                assemblyVersion.Major,
+                precision >= VersionOptions.VersionPrecision.Minor ? assemblyVersion.Minor : 0,
+                precision >= VersionOptions.VersionPrecision.Build ? assemblyVersion.Build : 0,
+                precision >= VersionOptions.VersionPrecision.Revision ? assemblyVersion.Revision : 0);
+        }
+
         return assemblyVersion;
     }
 

--- a/src/NerdBank.GitVersioning.Tests/VersionOracleTests.cs
+++ b/src/NerdBank.GitVersioning.Tests/VersionOracleTests.cs
@@ -259,6 +259,32 @@ public abstract class VersionOracleTests : RepoTestBase
         Assert.Matches(@"^2.3.1-g[a-f0-9]{7}$", oracle.ChocolateyPackageVersion);
     }
 
+    [Theory]
+    [InlineData("1.2.0.0", null, null)]
+    [InlineData("1.0.0.0", null, VersionOptions.VersionPrecision.Major)]
+    [InlineData("1.2.0.0", null, VersionOptions.VersionPrecision.Minor)]
+    [InlineData("1.2.1.0", null, VersionOptions.VersionPrecision.Build)]
+    [InlineData("2.3.4.0", "2.3.4", null)]
+    [InlineData("2.3.4.0", "2.3.4", VersionOptions.VersionPrecision.Minor)]
+    [InlineData("2.3.4.0", "2.3.4", VersionOptions.VersionPrecision.Build)]
+    [InlineData("2.3.4.0", "2.3.4.0", VersionOptions.VersionPrecision.Revision)]
+    public void CustomAssemblyVersion(string expectedAssemblyVersion, string prescribedAssemblyVersion, VersionOptions.VersionPrecision? precision)
+    {
+        this.InitializeSourceControl(withInitialCommit: false);
+        this.WriteVersionFile(new VersionOptions
+        {
+            Version = new SemanticVersion("1.2"),
+            AssemblyVersion = new VersionOptions.AssemblyVersionOptions
+            {
+                Version = prescribedAssemblyVersion is object ? new Version(prescribedAssemblyVersion) : null,
+                Precision = precision,
+            },
+        });
+
+        VersionOracle oracle = this.GetVersionOracle();
+        Assert.Equal(expectedAssemblyVersion, oracle.AssemblyVersion.ToString());
+    }
+
     [Fact]
     public void DefaultNuGetPackageVersionIsSemVer1PublicRelease()
     {

--- a/src/NerdBank.GitVersioning/VersionOracle.cs
+++ b/src/NerdBank.GitVersioning/VersionOracle.cs
@@ -444,14 +444,25 @@ namespace Nerdbank.GitVersioning
             // If there is no repo, "version" could have uninitialized components (-1).
             version = version.EnsureNonNegativeComponents();
 
-            var assemblyVersion = versionOptions?.AssemblyVersionOrDefault.Version ?? new System.Version(version.Major, version.Minor);
-            var precision = versionOptions?.AssemblyVersionOrDefault.PrecisionOrDefault;
+            Version assemblyVersion;
 
-            assemblyVersion = new System.Version(
-                assemblyVersion.Major,
-                precision >= VersionOptions.VersionPrecision.Minor ? assemblyVersion.Minor : 0,
-                precision >= VersionOptions.VersionPrecision.Build ? version.Build : 0,
-                precision >= VersionOptions.VersionPrecision.Revision ? version.Revision : 0);
+            if (versionOptions?.AssemblyVersion?.Version is not null)
+            {
+                // When specified explicitly, use the assembly version as the user defines it.
+                assemblyVersion = versionOptions.AssemblyVersion.Version;
+            }
+            else
+            {
+                // Otherwise consider precision to base the assembly version off of the main computed version.
+                VersionOptions.VersionPrecision precision = versionOptions?.AssemblyVersion?.Precision ?? VersionOptions.DefaultVersionPrecision;
+
+                assemblyVersion = new Version(
+                    version.Major,
+                    precision >= VersionOptions.VersionPrecision.Minor ? version.Minor : 0,
+                    precision >= VersionOptions.VersionPrecision.Build ? version.Build : 0,
+                    precision >= VersionOptions.VersionPrecision.Revision ? version.Revision : 0);
+            }
+
             return assemblyVersion.EnsureNonNegativeComponents(4);
         }
 

--- a/src/NerdBank.GitVersioning/version.schema.json
+++ b/src/NerdBank.GitVersioning/version.schema.json
@@ -44,8 +44,15 @@
             { "$ref": "#/definitions/twoToFourComponentVersion" },
             {
               "type": "object",
+              "additionalProperties": false,
               "properties": {
-                "version": { "$ref": "#/definitions/twoToFourComponentVersion" },
+                "version": { "$ref": "#/definitions/twoToFourComponentVersion" }
+              }
+            },
+            {
+              "type": "object",
+              "additionalProperties": false,
+              "properties": {
                 "precision": {
                   "type": "string",
                   "description": "Identifies the last component to be explicitly set in the version.",

